buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle.reaper;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A standalone process that will reap external services after a build dies.
+ *
+ * <h2>Input</h2>
+ * Since how to reap a given service is platform and service dependent, this tool
+ * operates on system commands to execute. It takes a single argument, a directory
+ * that will contain files with reaping commands. Each line in each file will be
+ * executed with {@link Runtime#getRuntime()#exec}.
+ *
+ * The main method will wait indefinitely on the parent process (Gradle) by
+ * reading from stdin. When Gradle shuts down, whether normally or abruptly, the
+ * pipe will be broken and read will return.
+ *
+ * The reaper will then iterate over the files in the configured directory,
+ * and execute the given commands. If any commands fail, a failure message is
+ * written to stderr. Otherwise, the input file will be deleted. If no inputs
+ * produced errors, the entire input directory is deleted upon completion of reaping.
+ */
+public class Reaper implements Closeable {
+
+    private Path inputDir;
+    private boolean failed;
+
+    private Reaper(Path inputDir) {
+        this.inputDir = inputDir;
+        this.failed = false;
+    }
+
+    public static void main(String[] args) throws Exception {
+        if (args.length != 1) {
+            System.err.println("Expected one argument.\nUsage: java -jar reaper.jar <DIR_OF_REAPING_COMMANDS>");
+            System.exit(1);
+        }
+        Path inputDir = Paths.get(args[0]);
```

Comment:
I think that is overkill. We don't do this for any other directories in the build.

Reply:
Maybe at least create the directory with such restricted permissions instead of relying on the default mask.  
This directory is different because we execute everything that gets written there, so it transforms the ability to write a file to that location into the ability to execute as the user running the daemon. 
Suppose an attacker gains a way to create files, this gives them the ability to execute them as a different user and do thinks like steal secrets available to Jenkins.

Reply:
That is an interesting case I had not thought of. However, Windows doesn't have permissions like this to set, so we wouldn't be covered in all cases. Additionally, the user that an attacker would likely be writing files as would the same user that gradle runs with, ie jenkins.

I think the best thing long term is to create a wire format to pass information to the reaper over stdin. However, I would like to do that as a followup because this currently works, and the case you described is an extreme edge case.

Reply:
Fun fact, the Gradle daemon itself isn't secured in any way. Anyone that communicate with it over a socket can issue arbitrary build actions that will run as the daemon user.

I think the approach I mentioned above for using stdin as the communications channel for which we tell the reaper about processes to kill is the better solution here vs relying on filesystem permissions. No one could hijack that as the reaper would effectively be "owned" by the daemon and only it could issue instructions. We can defer that for a later date though I think.